src/Components/Components/src/Routing/OptionalTypeRouteConstraint.cs

```diff
@@ -0,0 +1,42 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Microsoft.AspNetCore.Components.Routing
+{
+    /// <summary>
+    /// A route constraint that allows the value to be null or parseable as the specified
+    /// type.
+    /// </summary>
+    /// <typeparam name="T">The type to which the value must be parseable.</typeparam>
+    internal class OptionalTypeRouteConstraint<T> : RouteConstraint
```

Comment:
Why is the generic useful here? Is this just to be able to accept a `TryParseDelegate(string str, out T result);`?

Reply:
Correct. Do you have other ideas on how this should be done? We might able to refactor the duplicated code in the RouteConstraint...